### DISCUSSION TITLE
Proper way to create a FDataGrid object of curves with various lengths?

### DISCUSSION THREAD

**Discussion by hovinh** [GAA-UAM/scikit-fda#384] [Q&A] [answered]
Hi scikit-fda team,

I have checked the doc and it seems the underlying assumption when creating `FDataGrid` object is that `data_matrix` must have the same shape. I have a data set, whose half of them having approx 50 points and the other half approx 70 points. In order to apply group-wise alignment, I would like to create a single `FDataGrid` object. How could I do that?

**Attempt 1**: create `data_matrix` with different length throws error, even when I don't clarify `grid_points`.
Code:
```python
import skfda

grid_points = [[0, 0.2, 0.5, 0.9, 1], [0, 0.2, 0.5, 0.9]]
data_matrix = np.array([
    [0, 0.2, 0.5, 0.9, 1],     # First observation
    [0, 0.04, 0.25, 0.81],  # Second observation
])

fd = skfda.FDataGrid(
    data_matrix=data_matrix,
    grid_points=grid_points,
)
```
Output:
```
 /opt/conda/envs/fda_py37/lib/python3.7/site-packages/ipykernel_launcher.py:5: VisibleDeprecationWarning:Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-52-720e51573dae> in <module>
      8 fd = skfda.FDataGrid(
      9     data_matrix=data_matrix,
---> 10     grid_points=grid_points,
     11 )
     12 

/opt/conda/envs/fda_py37/lib/python3.7/site-packages/skfda/representation/grid.py in __init__(self, data_matrix, grid_points, sample_points, domain_range, dataset_label, dataset_name, argument_names, coordinate_names, sample_names, axes_labels, extrapolation, interpolation)
    159             grid_points = sample_points
    160 
--> 161         self.data_matrix = _int_to_real(np.atleast_2d(data_matrix))
    162 
    163         if grid_points is None:

/opt/conda/envs/fda_py37/lib/python3.7/site-packages/skfda/_utils/_utils.py in _int_to_real(array)
    654 def _int_to_real(array: np.ndarray) -> np.ndarray:
    655     """Convert integer arrays to floating point."""
--> 656     return array + 0.0
    657 
    658 

TypeError: can only concatenate list (not "float") to list
```

**Attempt 2**: Create a FDataGrid for each trace, then l need a function to merge them. Do you have such thing?

**Attempt 3**: Resample each trace to a median length. This is doable but I'm afraid certain information could be lost that potentially affect the warping result.

Could you please advise on what is the proper way to do it?

Thank you,
Vinh

**Comment by vnmabus** [comment, marked as answer]
What do you mean by different lengths? As far as I know, that could mean two different things:
- "Partial" curves where some fragments of the curve are missing. This could probably be represented using NaN for missing values, but you should expect most methods to fail, as that case (currently) has not been taken into consideration.
- Curves in the same interval but sampled at different points, such as longitudinal data. We should create a new `FData` subclass to handle this kind of data, but currently we are not prioritizing that. A possible workaround is to express your data in a common grid (by interpolation) or project it to a common basis. Once you have both with the exact same representation you can use `skfda.concatenate` to join them. Issue #325 discuss a possible future way to implement that as part of a pipeline.

Which one is your case?

**Reply by hovinh** [reply]
My data has the **same sampling rate**, but **different interval**  - some last longer than the others. For example, it could be athletes' heart rate to complete a 5-km distance; each person completes the race at different time interval, but the heart rate has the same sampling rate.

**Reply by vnmabus** [reply]
We don't have a way to work with that kind of data. Most of the literature in FDA deals only with functions in the same domain. I guess that you could scale the curve to the same "domain", maybe keeping the scale factor as an additional scalar feature, but in your example that wouldn't make much sense (data couldn't be properly aligned and most FDA methods would perform poorly). What kind of analysis did you want to do with that data?

**Reply by hovinh** [reply]
I want to employ warping function as an alternative for vanilla DTW on time-series domain, so that I can establish a 3-sigma threshold, or so-called smart band, to detect any new timeseries that deviates from it.

The procedure can be simplified as follows:
- Training: You have a golden dataset of time series, group-wise align them to compute the mean trace and sigma point by point. From there, you can establish a smart band defined by `mean +- 3sigma`.
- Inference: new timeseries to be warped to established mean trace, then I can devise a distance metric to indicate whether is is an anomaly trace or not.

As above, warping plays an important role in both training and inference. Empirical experiments show that vanilla DTW produced unexpected mapping, and its variants do not perform well on all cases. So far `ElasticRegistration` does pretty well on my prepared dataset, and faster, so I would like to add it as default warping method. The only constraint is the same-length input as said.

**Reply by vnmabus** [reply]
So, how are you computing the mean and sigma for missing points? Do you just ignore the missing values?

In an example such as the heart rate one, longer curves will have more peaks and valleys that shorter ones. What do you expect `ElasticRegistration` to do in this case? It works by computing a global template and aligning everything towards it, so how many peaks/valleys should the template have? What will it do with longer/shorter curves?

**Reply by hovinh** [reply]
During training, after traces have been warped to the same template, we can ensure each trace has the same number of data points. I can compute the mean and sigma for each index. For example, if our trace has 20 data points, I will have 20 means and sigmas. Therefore, there is no such "missing points" issue.

You are correct in the heart rate example. I could not share, but we have domain knowledge that our dataset has the same peak/valley amount predetermined, regardless of trace length. 

By observing the warping function plot, I judge whether the outcome is expected or not: we know despite the difference in length, the shifted amount is local. Most of the time, I would expect the warping function is as close to the diagonal line as possible. Speaking of which, based on the literature, it said the warping function's assumption is that the magnitude is minimal between warped points. Is it true for me to say `ElasticRegistration` has a local constraint (prefer mapping to values in proximity) in its optimization target?

**Reply by vnmabus** [reply]
Honestly, I am not an expert in `ElasticRegistration`. It has a regularization parameter, but if you set that to 0 I do not think that it has a local constraint. However it registers each curve toward the template by using the derivatives (via the SRVF), so it tends to align the shape, ignoring potential differences in amplitude, if I am not mistaken. If you want more precise information I guess you could contact @jdtuck, who is the main maintainer of the elastic registration packages for several programming languages and also has several articles about the SRVF framework.

**Reply by jdtuck** [reply]
There is no local constraint and it is aligning shape so peaks nearby are going to be matched. At its core it is a minimal energy minimization using Dynamic Programming. From what I have read above you are trying to do partial curve matching which is a much harder problem.

**Reply by hovinh** [reply]
Thank you for your comment, vnmabus and jdtuck. Since we are on the topic of warping function, I have these follow-up comments/questions, hope you don't mind:
- Heart rate was not a good example because I would like to look for global alignment between two traces, not their sub-trace (partial curve matching). A better one is the typical handwriting example; each person writes a word at different speed, hence the recording of (x, y, z) of their handwriting should come in different length. 
-  Local constraint: I use this term to imply the index mapping between the 2 traces should not be too distant, quite similar to Sakoe-Chiba DTW variant. However, in the context of cost matrix (or warping function plot), what  Sakoe-Chiba DTW does is *limiting the search space of optimal path*. May I ask what is the corresponding interpretation for SRSF?
![image](https://user-images.githubusercontent.com/7614509/139678572-577f613d-4089-4839-b16f-bac36ffa9d9f.png)
- The non-constrained `ElasticRegistration`  work well for me, but I'm curious if I want to tweak `penalty_term`, what are the recommended parameter value that I can try out?
- Is *landmark registration* a divide-and-conquer version of `ElasticRegistration`, that applies in each interval defined by landmark pairs, or the warping is computed based on equation (1) from [Functional Data Analysis of Amplitude and Phase Variation](https://arxiv.org/pdf/1512.03216.pdf)?
![image](https://user-images.githubusercontent.com/7614509/139681898-0e4cf736-4583-4753-9f94-eb47bfc37174.png)
- I understand that the "same-length" condition is chosen to reflect the FDA literature, but is it possible to show me a way to hack the existing `ElastingRegistration` so that it can accept two traces in different length and outputs the same length?

Thank you :D

**Reply by jdtuck** [reply]
The penalty term in the elastic registration limits how far the warping function can be from identity (diagonal line). The larger the number the more it is constrained. There is no way to tweak the code as you are at that point violating the theory. The theory is that each of the curves are observed on the same domain [0,T]. If you wish to use this, I would interpolate the curves to the same domain.

**Reply by vnmabus** [reply]
Landmark registration has nothing to do with (Fisher Rao) elastic registration. It is only "elastic" in the sense that it does not only use rigid transformations, but allows time deformations too. It construct a warping by placing the corresponding landmark points at the same position in system time and interpolating the values between them.

**Reply by hovinh** [reply]
Thank you :))

**Reply by hovinh** [reply]
Hi,

I would like to resume this conversation. I have tried out pairwise alignment and notice that the input trace can have different length from the template. After transformation, the trace takes the original length, not template length. Not sure if this is a bug or not?

Example 1:
```python
import matplotlib.colors as clr
import matplotlib.pyplot as plt
import numpy as np
import skfda

from skfda.preprocessing.registration import ElasticRegistration, invert_warping
from skfda.datasets import make_multimodal_samples

# Samples with modes in 1/3 and 2/3
fd = make_multimodal_samples(n_samples=2, modes_location=[1 / 3, 2 / 3],
                             random_state=1, start=0, mode_std=.01)

fig = fd.plot()
fig.axes[0].legend(['$f$', '$g$'])
plt.show()

def crop_fd_data(fd, size):
    data_matrix = fd.data_matrix[:, :size, :]
    grid_points = fd.grid_points[0][:size]
    return skfda.FDataGrid(
        data_matrix=data_matrix,
        grid_points=grid_points
    )

f = crop_fd_data(fd[0], 60) 
g = fd[1]
elastic_registration = ElasticRegistration(template=g)


# Aligns f to g
f_align = elastic_registration.fit_transform(f)

fig = f.plot()
fig = g.plot(fig=fig)
f_align.plot(fig=fig, color='r', linestyle='--')

# Legend
fig.axes[0].legend(['$f$', '$g$', '$f \\circ \\gamma $'])
plt.show()
```
Output:
![image](https://user-images.githubusercontent.com/7614509/142342910-c31b13fd-1640-43bf-afa7-362fda3c2b12.png)
![image](https://user-images.githubusercontent.com/7614509/142342924-5762db35-ea04-41e1-ac7f-eeb00ee6b7e1.png)


Example 2:
```python
def crop_fd_data(fd, size):
    data_matrix = fd.data_matrix[:, :size, :]
    grid_points = fd.grid_points[0][:size]
    return skfda.FDataGrid(
        data_matrix=data_matrix,
        grid_points=grid_points
    )

f = fd[0] 
g = crop_fd_data(fd[1], 60)
elastic_registration = ElasticRegistration(template=g)


# Aligns f to g
f_align = elastic_registration.fit_transform(f)

fig = f.plot()
fig = g.plot(fig=fig)
f_align.plot(fig=fig, color='r', linestyle='--')

# Legend
fig.axes[0].legend(['$f$', '$g$', '$f \\circ \\gamma $'])
plt.show()
```
Output:
![image](https://user-images.githubusercontent.com/7614509/142343073-48996ff1-7def-4388-be64-16122c13c7a6.png)

**Reply by vnmabus** [reply]
I don't think that it ever makes sense to pass a template with different grid points than the curves. I think we should check for that in fit and predict, to prevent confusion.

**Reply by jdtuck** [reply]
I agree with that. It goes against the theory I the package. Partial curve matching is much harder and more complex problem.
